src/index.ts

```diff
@@ -14,637 +14,199 @@
-      this.importLibrary(library)
-    );
-    // ensure at least one library, to kick off loading...
-    if (!libraryPromises.length) {
-      libraryPromises.push(this.importLibrary("core"));
-    }
-    Promise.all(libraryPromises).then(
-      () => this.callback(),
-      (error) => {
-        const event = new ErrorEvent("error", { error }); // for backwards compat
-        this.loadErrorCallback(event);
-      }
-    );
+  const name = libraryName as keyof APILibraryMap;
+  if (!libraries_[name]) {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    libraries_[name] = (await google.maps.importLibrary(name)) as any;
   }
 
-  /**
-   * Reset the loader state.
-   */
-  private reset(): void {
-    this.deleteScript();
-    this.done = false;
-    this.loading = false;
-    this.errors = [];
-    this.onerrorEvent = null;
-  }
+  return libraries_[name] as APILibraryMap[keyof APILibraryMap];
+}
 
-  private resetIfRetryingFailed(): void {
-    if (this.failed) {
-      this.reset();
+/**
+ * The bootstrap function makes sure that a usable version of the
+ * google.maps.importLibrary function exists and loading of the maps API is
+ * started.
+ */
+function bootstrap(options: APIOptions) {
+  // check for maps JS already being loaded
+  const scriptEl = document.querySelector(
+    'script[src*="maps.googleapis.com/maps/api/js"]'
+  ) as HTMLScriptElement | null;
+
+  if (!scriptEl) {
+    // bootstrapInternal will setup the google.maps.importLibrary function to
+    // start loading the Maps API when needed.
+    bootstrapInternal(options);
+    return;
+  }
+
+  // Check the parameters to be compatible with `options` passed to this function:
+  // - a different api key or different version will log an error
+  // - different language or region will trigger warning in dev-mode
+  // - in any other case we just log an informational message in dev-mode
+  const scriptOptions = getOptionsFromURL(scriptEl.src);
+  if (scriptOptions.key !== options.key || scriptOptions.v !== options.v) {
+    logError(ERR_KEY_VERSION_MISMATCH(options, scriptOptions));
+  } else if (
+    scriptOptions.language !== options.language ||
+    scriptOptions.region !== options.region
+  ) {
+    logDevWarning(WARN_LANGUAGE_REGION_NOT_COMPATIBLE(options, scriptOptions));
+  } else {
+    logDevNotice(INFO_LOADED_WITHOUT_OPTIONS);
+  }
+
+  // there is a script tag, and importLibrary is defined, so we're all set
+  if (window.google?.maps?.importLibrary) return;
+
+  // maps JS is currently being loaded, but wasn't loaded with the dynamic
+  // import snipped. In this case, the importLibrary bootstrap function has to
+  // be provided
+  const scriptLoadedPromise = new Promise<void>((resolve, reject) => {
+    // FIXME: it would probably be more precise to intercept (override and call
+    //   the original) the callback function specified in
+    //   `scriptOptions.callback`, but the load event should be sufficient.
```

Comment:
actually, the load event is not safe to use (we can defer calling `callback` after subsequent script loads happen)